File: src/gui/SettingsPane.java
package gui;

import gui.components.NumberTextFieldSetting;
import gui.components.SliderSetting;
import javafx.geometry.HPos;
import javafx.geometry.Insets;
import javafx.scene.layout.*;

public class SettingsPane extends GridPane
{
    private static final int COL_INDEX_LABEL = 0;
    private static final int COL_INDEX_TEXTFIELD = 1;
    private static final int COL_INDEX_SLIDER = 2;

    private static final String SERVER_PORT_LABEL = "Server Port:";
    private static final double SERVER_PORT_MAX = 65535;
    private static final double SERVER_PORT_MIN = 1;
    private static final double SERVER_PORT_DEFAULT = 7005;

    private static final String NETWORK_CAPACITY_LABEL = "Capacity:";
    private static final double NETWORK_CAPACITY_MAX = 10000;
    private static final double NETWORK_CAPACITY_MIN = 0;
    private static final double NETWORK_CAPACITY_DEFAULT = 3000;

    private static final String PACKET_DROP_FUN_LABEL = "Packet Drop Function:";
    private static final double PACKET_DROP_FUN_MAX = 6;
    private static final double PACKET_DROP_FUN_MIN = 0;
    private static final double PACKET_DROP_FUN_DEFAULT = 4;

    private static final String NOISE_LABEL = "Noise:";
    private static final double NOISE_MAX = 1;
    private static final double NOISE_MIN = 0;
    private static final double NOISE_DEFAULT = 0.01;

    private static final String LATENCY_LABEL = "Latency:";
    private static final double LATENCY_MAX = 5000;
    private static final double LATENCY_MIN = 0;
    private static final double LATENCY_DEFAULT = 250;

    private static final String JITTER_LABEL = "Jitter:";
    private static final double JITTER_MAX = 5000;
    private static final double JITTER_MIN = 0;
    private static final double JITTER_DEFAULT = 250;

    private int nextRow;

    public SettingsPane()
    {
        super();

        // initialize instance data
        nextRow = 0;

        // configure aesthetic properties
        setPadding(new Insets(Dimens.KEYLINE_SMALL));
        setHgap(Dimens.KEYLINE_SMALL);
        setVgap(Dimens.KEYLINE_SMALL);

        // configure grid constraints
        ColumnConstraints column1 = new ColumnConstraints();
        column1.setHalignment(HPos.RIGHT);
        getColumnConstraints().add(0,column1);

        ColumnConstraints column2 = new ColumnConstraints();
        getColumnConstraints().add(1,column2);

        ColumnConstraints column3 = new ColumnConstraints();
        column3.setFillWidth(true);
        column3.setHgrow(Priority.ALWAYS);
        getColumnConstraints().add(2,column3);

        // serverPortControl
        NumberTextFieldSetting serverPortControl = new NumberTextFieldSetting(false);
        serverPortControl.getLabel().setText(SERVER_PORT_LABEL);
        serverPortControl.getNumberTextField().setMax(SERVER_PORT_MAX);
        serverPortControl.getNumberTextField().setMin(SERVER_PORT_MIN);
        serverPortControl.getNumberTextField().setValue(SERVER_PORT_DEFAULT);
        addNumberTextFieldSetting(serverPortControl);

        // packetDropFunctionControl
        SliderSetting networkCapacityControl = new SliderSetting(false);
        networkCapacityControl.getLabel().setText(NETWORK_CAPACITY_LABEL);
        networkCapacityControl.getSlider().setMax(NETWORK_CAPACITY_MAX);
        networkCapacityControl.getSlider().setMin(NETWORK_CAPACITY_MIN);
        networkCapacityControl.getSlider().setValue(NETWORK_CAPACITY_DEFAULT);
        addSliderSetting(networkCapacityControl);

        // packetDropFunctionControl
        SliderSetting packetDropFunctionControl = new SliderSetting(true);
        packetDropFunctionControl.getLabel().setText(PACKET_DROP_FUN_LABEL);
        packetDropFunctionControl.getSlider().setMax(PACKET_DROP_FUN_MAX);
        packetDropFunctionControl.getSlider().setMin(PACKET_DROP_FUN_MIN);
        packetDropFunctionControl.getSlider().setValue(PACKET_DROP_FUN_DEFAULT);
        addSliderSetting(packetDropFunctionControl);

        // noiseControl
        SliderSetting noiseControl = new SliderSetting(true);
        noiseControl.getLabel().setText(NOISE_LABEL);
        noiseControl.getSlider().setMax(NOISE_MAX);
        noiseControl.getSlider().setMin(NOISE_MIN);
        noiseControl.getSlider().setValue(NOISE_DEFAULT);
        addSliderSetting(noiseControl);

        // latencyControl
        SliderSetting latencyControl = new SliderSetting(false);
        latencyControl.getLabel().setText(LATENCY_LABEL);
        latencyControl.getSlider().setMax(LATENCY_MAX);
        latencyControl.getSlider().setMin(LATENCY_MIN);
        latencyControl.getSlider().setValue(LATENCY_DEFAULT);
        addSliderSetting(latencyControl);

        // jitterControl
        SliderSetting jitterControl = new SliderSetting(false);
        jitterControl.getLabel().setText(JITTER_LABEL);
        jitterControl.getSlider().setMax(JITTER_MAX);
        jitterControl.getSlider().setMin(JITTER_MIN);
        jitterControl.getSlider().setValue(JITTER_DEFAULT);
        addSliderSetting(jitterControl);
    }

    private void addSliderSetting(SliderSetting newSliderSetting)
    {
        add(newSliderSetting.getLabel(),COL_INDEX_LABEL,nextRow);
        add(newSliderSetting.getNumberTextField(),COL_INDEX_TEXTFIELD,nextRow);
        add(newSliderSetting.getSlider(),COL_INDEX_SLIDER,nextRow);
        nextRow++;
    }

    private void addNumberTextFieldSetting(NumberTextFieldSetting newNumberTextField)
    {
        add(newNumberTextField.getLabel(),COL_INDEX_LABEL,nextRow);
        add(newNumberTextField.getNumberTextField(),COL_INDEX_TEXTFIELD,nextRow);
        nextRow++;
    }
}

File: netsim/src/gui/SettingsPane.kt
package gui

import javafx.beans.property.*
import javafx.geometry.HPos
import javafx.geometry.Insets
import javafx.scene.control.Label
import javafx.scene.control.Slider
import javafx.scene.layout.ColumnConstraints
import javafx.scene.layout.GridPane
import javafx.scene.layout.Priority
import parse

internal class SettingsPane:GridPane()
{
    val serverPort:IntegerProperty = SimpleIntegerProperty(0)
    val networkCapacity:IntegerProperty = SimpleIntegerProperty(0)
    val packetDropFunction:DoubleProperty = SimpleDoubleProperty(0.0)
    val noise:DoubleProperty = SimpleDoubleProperty(0.0)
    val jitter:LongProperty = SimpleLongProperty(0)
    val latency:LongProperty = SimpleLongProperty(0)

    private val COL_INDEX_LABEL = 0
    private val COL_INDEX_TEXTFIELD = 1
    private val COL_INDEX_SLIDER = 2

    private val SERVER_PORT_LABEL = "Server Port:"
    private val SERVER_PORT_MAX = 65535.0
    private val SERVER_PORT_MIN = 1.0
    private val SERVER_PORT_DEFAULT = 7005.0

    private val NETWORK_CAPACITY_LABEL = "Capacity:"
    private val NETWORK_CAPACITY_MAX = 10000.0
    private val NETWORK_CAPACITY_MIN = 0.0
    private val NETWORK_CAPACITY_DEFAULT = 3000.0

    private val PACKET_DROP_FUN_LABEL = "Packet Drop Function:"
    private val PACKET_DROP_FUN_MAX = 6.0
    private val PACKET_DROP_FUN_MIN = 0.0
    private val PACKET_DROP_FUN_DEFAULT = 4.0

    private val NOISE_LABEL = "Noise:"
    private val NOISE_MAX = 1.0
    private val NOISE_MIN = 0.0
    private val NOISE_DEFAULT = 0.01

    private val LATENCY_LABEL = "Latency:"
    private val LATENCY_MAX = 5000.0
    private val LATENCY_MIN = 0.0
    private val LATENCY_DEFAULT = 250.0

    private val JITTER_LABEL = "Jitter:"
    private val JITTER_MAX = 5000.0
    private val JITTER_MIN = 0.0
    private val JITTER_DEFAULT = 250.0

    private var nextRow:Int = 0

    init
    {
        configureLayout()
        addChildNodes()
    }

    private fun configureLayout()
    {
        // configure aesthetic properties
        padding = Insets(Dimens.KEYLINE_SMALL.toDouble())
        hgap = Dimens.KEYLINE_SMALL.toDouble()
        vgap = Dimens.KEYLINE_SMALL.toDouble()

        // configure grid constraints
        val column1 = ColumnConstraints()
        column1.halignment = HPos.RIGHT
        columnConstraints.add(0,column1)

        val column2 = ColumnConstraints()
        columnConstraints.add(1,column2)

        val column3 = ColumnConstraints()
        column3.isFillWidth = true
        column3.hgrow = Priority.ALWAYS
        columnConstraints.add(2,column3)
    }

    private fun addChildNodes()
    {
        // serverPortControl
        val serverPortControl = SliderSetting(false)
        serverPortControl.label.text = SERVER_PORT_LABEL
        serverPortControl.slider.max = SERVER_PORT_MAX
        serverPortControl.slider.min = SERVER_PORT_MIN
        serverPortControl.slider.value = SERVER_PORT_DEFAULT
        serverPortControl.slider.valueProperty().addListener{
            value,oldValue,newValue ->
            serverPort.value = Math.round(newValue.toFloat())
        }

        // packetDropFunctionControl
        val networkCapacityControl = SliderSetting(false)
        networkCapacityControl.label.text = NETWORK_CAPACITY_LABEL
        networkCapacityControl.slider.max = NETWORK_CAPACITY_MAX
        networkCapacityControl.slider.min = NETWORK_CAPACITY_MIN
        networkCapacityControl.slider.value = NETWORK_CAPACITY_DEFAULT
        networkCapacityControl.slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                networkCapacity.value = Math.round(newValue.toFloat())
            })

        // packetDropFunctionControl
        val packetDropFunctionControl = SliderSetting(true)
        packetDropFunctionControl.label.text = PACKET_DROP_FUN_LABEL
        packetDropFunctionControl.slider.max = PACKET_DROP_FUN_MAX
        packetDropFunctionControl.slider.min = PACKET_DROP_FUN_MIN
        packetDropFunctionControl.slider.value = PACKET_DROP_FUN_DEFAULT
        packetDropFunctionControl.slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                packetDropFunction.value = newValue.toDouble()
            })

        // noiseControl
        val noiseControl = SliderSetting(true)
        noiseControl.label.text = NOISE_LABEL
        noiseControl.slider.max = NOISE_MAX
        noiseControl.slider.min = NOISE_MIN
        noiseControl.slider.value = NOISE_DEFAULT
        noiseControl.slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                noise.value = newValue.toDouble()
            })

        // latencyControl
        val latencyControl = SliderSetting(false)
        latencyControl.label.text = LATENCY_LABEL
        latencyControl.slider.max = LATENCY_MAX
        latencyControl.slider.min = LATENCY_MIN
        latencyControl.slider.value = LATENCY_DEFAULT
        latencyControl.slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                latency.value = Math.round(newValue.toDouble())
            })

        // jitterControl
        val jitterControl = SliderSetting(false)
        jitterControl.label.text = JITTER_LABEL
        jitterControl.slider.max = JITTER_MAX
        jitterControl.slider.min = JITTER_MIN
        jitterControl.slider.value = JITTER_DEFAULT
        jitterControl.slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                jitter.value = Math.round(newValue.toDouble())
            })

        // add nodes to layout
        addNumberTextFieldSetting(serverPortControl)
        addSliderSetting(networkCapacityControl)
        addSliderSetting(packetDropFunctionControl)
        addSliderSetting(noiseControl)
        addSliderSetting(latencyControl)
        addSliderSetting(jitterControl)
    }

    private fun addSliderSetting(newSliderSetting:SliderSetting)
    {
        add(newSliderSetting.label,COL_INDEX_LABEL,nextRow)
        add(newSliderSetting.doubleTextField,COL_INDEX_TEXTFIELD,nextRow)
        add(newSliderSetting.slider,COL_INDEX_SLIDER,nextRow)
        nextRow++
    }

    private fun addNumberTextFieldSetting(newSliderSetting:SliderSetting)
    {
        add(newSliderSetting.label,COL_INDEX_LABEL,nextRow)
        add(newSliderSetting.doubleTextField,COL_INDEX_TEXTFIELD,nextRow)
        nextRow++
    }
}

private class SliderSetting(private val allowDecimalNumbers:Boolean)
{
    val label:Label = Label()
    var doubleTextField:DoubleTextField = DoubleTextField()
    var slider:Slider = Slider()

    init
    {
        // configure numberTextField so when its value changes, it updates slider
        doubleTextField.textProperty().addListener(
            {
                value,oldValue,newValue ->
                slider.value = Double.parse(newValue)
                doubleTextField.text = if (allowDecimalNumbers)
                    slider.value.toString()
                else Math.round(slider.value).toString()
            })

        // configure slider so when its value changes, it updates numberTextField
        slider.maxWidth = java.lang.Double.MAX_VALUE
        slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                doubleTextField.text = newValue.toString()
            })
    }

    fun valueProperty():DoubleProperty
    {
        return slider.valueProperty()
    }
}
